reconcile/external_resources/state.py

```diff
@@ -0,0 +1,209 @@
+import logging
+from collections.abc import Mapping
+from datetime import datetime, timezone
+from enum import Enum
+from typing import Any
+
+import boto3
+from pydantic import BaseModel
+
+from reconcile.external_resources.model import (
+    ExternalResourceKey,
+    ExternalResourceModuleConfiguration,
+    Reconciliation,
+)
+
+DATE_FORMAT = "%Y-%m-%dT%H:%M:%SZ"
+
+
+class StateNotFoundError(Exception):
+    pass
+
+
+class ReconcileStatus(str, Enum):
+    SUCCESS: str = "SUCCESS"
+    ERROR: str = "ERROR"
+    IN_PROGRESS: str = "IN_PROGRESS"
+    NOT_EXISTS: str = "NOT_EXISTS"
+
+
+class ResourceStatus(str, Enum):
+    CREATED: str = "CREATED"
+    DELETED: str = "DELETED"
+    ABANDONED: str = "ABANDONED"
+    NOT_EXISTS: str = "NOT_EXISTS"
+    IN_PROGRESS: str = "IN_PROGRESS"
+    DELETE_IN_PROGRESS: str = "DELETE_IN_PROGRESS"
+    ERROR: str = "ERROR"
+
+
+class ExternalResourceState(BaseModel):
+    key: ExternalResourceKey
+    ts: datetime
+    resource_status: ResourceStatus
+    resource_hash: str = ""
+    reconciliation: Reconciliation
+    reconciliation_errors: int = 0
+
+
+class DynamoDBStateAdapater:
+    KEY_PROVISION_PROVIDER = "key.provision_provider"
+    KEY_PROVISIONER_NAME = "key.provisioner_name"
+    KEY_PROVIDER = "key.provider"
+    KEY_IDENTIFIER = "key.identifier"
+    RECONC_RESOURCE_HASH = "reconc.resource_hash"
+    RECONC_MODCONF_IMAGE = "reconc.modconf.image"
+    RECONC_MODCONF_VERSION = "reconc.modconf.version"
+    RECONC_MODCONF_DRIFT_MINS = "reconc.modconf.drift_mins"
+    RECONC_MODCONF_TIMEOUT_MINS = "reconc.modconf.timeout_mins"
+    RECONC_INPUT = "reconc.input"
+    RECONC_ACTION = "reconc.action"
+    RECONC_ERRORS = "reconciliation_errors"
+    RESOURCE_KEY = "resource_key"
+    RESOURCE_STATUS = "resource_status"
+    RESOURCE_HASH = "resource_hash"
+    TIMESTAMP = "ts"
```

Comment:
I'm not clear about the dynamodb schema and index, I checked in dev account, the PK is `resource_key`, and there is a GSI `index_resources` also using the same PK `resource_key`, seems not needed. Can you clarify dynamodb schema and usages?

Reply:
btw, typo `Adapater` -> `Adapter`

Reply:
You are right. 
I wanted to get some fields of each record to make the initial Desired-Current comparison and I thought an index was required for that, but it is not. I changed the code to use the table with the required fields as projected attributes instead of an index. 

I also changed the structure to use document types and simplified it a bit. PTAL.